opm/simulators/flow/ExtraConvergenceOutputThread.cpp

```diff
@@ -0,0 +1,321 @@
+{
+    if (config.want(ConvergenceOutputConfiguration::Option::Iterations)) {
+        this->infoIter_.emplace
+            (std::filesystem::path { outputDir } /
+             std::filesystem::path { baseName }.concat(".INFOITER"));
+    }
+}
+
+void
+Opm::ConvergenceOutputThread::Impl::
+write(const std::vector<ConvergenceReportQueue::OutputRequest>& requests)
+{
+    assert (! requests.empty() &&
+            "Internal logic error in forming convergence output request");
+
+    this->writeIterInfo(requests);
+}
+
+void
+Opm::ConvergenceOutputThread::Impl::
+writeIterInfo(const std::vector<ConvergenceReportQueue::OutputRequest>& requests)
+{
+    if (! this->infoIter_.has_value()) {
+        return;
+    }
+
+    if (! this->haveOutputIterHeader_) {
+        this->colSize_ =
+            writeConvergenceHeader(this->infoIter_.value(),
+                                   this->getPhaseName_,
+                                   requests.front());
+        this->haveOutputIterHeader_ = true;
+    }
+
+    for (const auto& request : requests) {
+        writeConvergenceRequest(this->infoIter_.value(),
+                                this->convertTime_,
+                                this->colSize_,
+                                request);
+
+        if (request.reports.empty()) {
+            this->finalRequestWritten_ = true;
+            break;
+        }
+    }
+
+    this->infoIter_.value().flush();
+}
+
+// ===========================================================================
+// Public Interface Below Separator
+// ===========================================================================
+
+void Opm::ConvergenceReportQueue::enqueue(std::vector<OutputRequest>&& requests)
+{
+    // Signal output thread if we're going from "no work" to "some work".
+    // We don't need to signal if we're going from "some work" to "more
+    // work".
+    const auto must_notify = this->requests_.empty();
+
+    {
+        std::lock_guard<std::mutex> guard{ this->mtx_ };
+        this->requests_.insert(this->requests_.end(),
+                               std::make_move_iterator(requests.begin()),
+                               std::make_move_iterator(requests.end()));
+    }
+
+    if (must_notify) {
+        this->cv_.notify_one();
+    }
+}
+
+void Opm::ConvergenceReportQueue::signalLastOutputRequest()
+{
+    // Empty request signals end of production.
+    this->enqueue(std::vector<OutputRequest>(1));
+}
+
+// ---------------------------------------------------------------------------
+
+Opm::ConvergenceOutputThread::
+ConvergenceOutputThread(std::string_view               outputDir,
+                        std::string_view               baseName,
+                        ComponentToPhaseName           getPhaseName,
+                        ConvertToTimeUnits             convertTime,
+                        ConvergenceOutputConfiguration config,
+                        ConvergenceReportQueue&        queue)
+    : pImpl_{ new Impl {
```

Comment:
make_unique

Reply:
> `make_unique<>()`

Hm, interesting.  I've not been successful using `make_unique<>()` in this particular situation before (i.e., "pointer to implementation") so I did not entertain the possibility that the better alternative would actually work now.  It does, however, so I've pushed an update to that effect.  I'll switch other "pimpl" instances to using `make_unique<>()` if I'm updating the surrounding code.